### PR TITLE
History page

### PR DESCRIPTION
**Pull request by ajnye**
## GitHub Issues addressed

- This PR closes nothing.

## What I did

- Created a recently viewed history page for professor pages inside a specific course. This modal is accessible from the navbar and landing.

## Screenshots

- Before
Screen Shot 2023-04-02 at 2 46 35 PM
Screen Shot 2023-04-02 at 2 46 46 PM
- After
Screen Shot 2023-04-02 at 2 47 19 PM
Screen Shot 2023-04-02 at 2 48 10 PM
Screen Shot 2023-04-02 at 2 48 25 PM

## Testing

- A brief explanation of tests done/written or how reviewers can test your work
Run the branch locally, look at the browse and landing pages and press it to take a look at recently viewed. Also click into course and professors to fill up the recently viewed page. It should display the most recently used ones at the top, not have any duplicates, and only display the last 10 pages visited. The modal only records the pages with the reviews themselves.

## Questions/Discussions/Notes

- How does the recently viewed page look? Are there any UI spacing issues when changing sizes?

### CODE REVIEW

**Comment by alex-shen1** [comment]
@ajnye Pylint is broken

**Review by Ann924 — COMMENTED** [review]
Everything else looks good! The format of the history panel and the new badge work on my display, even when zooming in/out and changing the window size.

**Comment by obs4qqf** [comment]
Nice job implementing this feature! I think the current feature sizing is perfect, it feels very easy to use for me. Here are some additional thoughts below:

- I tested out what happens when I hit the 404 page, and I'm getting a bug where invalid links are stored, but the courses/titles displayed are still normal. This seems to cause a mismatch between the courses on the history page and the link attached to it
	- (I activated 404 pages locally by changing DEBUG=0 in .env and changing a line in the `docker-compose.yml` file to `python3 manage.py runserver 0.0.0.0:8000 --insecure`)
- When there is no courses/instructors viewed yet, it might be good to display a message like "No courses viewed yet" for clarity
- Another idea I had that might be out of scope is it might be good for the user to be able to clear their history page through some sort of button (though this might not be a top concern)

File: tcf_core/settings/record_middleware.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            # Converts string representation of list into list object
+            previous_paths = ast.literal_eval(previous_paths)
+
+            previous_paths_titles = request.COOKIES['previous_paths_titles']
+            # Converts string representation of list into list object
+            previous_paths_titles = ast.literal_eval(previous_paths_titles)
+
+            # Keeps top 10 items in list
+            # Only keeps 9 here since 1 is inserted immediately after
+            if len(previous_paths) > 9:
```
[the remark this answers is not in the document]
Comment (ajnye) [reply tcf_core/settings/record_middleware.py]: Fixed it, thanks for the review!

**Comment by ajnye** [comment]
Fixed request to display message for no courses viewed, can not currently fix 404 page issue, but it is low priority.

File: tcf_website/views/index.py [hunk 2]
```diff
@@ -23,7 +23,7 @@ def index(request):
                        'FAQs': faqs,
                        'visited': request.session.get('visited', False)})
     request.session['visited'] = True
-
+    print("test test tset")
```
Comment (jackliu612) [inline tcf_website/views/index.py:26]: please remove haha

File: tcf_core/context_processors.py [hunk 3]
```diff
@@ -10,3 +12,36 @@ def base(request):
...
+
+    previous_paths_titles = [title[:80] + '...' for title in previous_paths_titles]
+
+    previous_paths_and_titles = None
+    if len(previous_paths) > 0 and len(previous_paths_titles) > 0:
+        previous_paths_and_titles = zip(previous_paths, previous_paths_titles)
+
+    return {
+        'previous_paths': previous_paths,
+        'count': count,
```
Comment (jackliu612) [inline tcf_core/context_processors.py:44]: I don't see where this `count` variable is being used, is it necessary? Is it always the same as `len(previous_paths)`? Also is `previous_paths` and `previous_path_titles` always going to be the same length?

**Review by jackliu612 — APPROVED** [review]

**Pull request merged**